### ISSUE TITLE
[RFC][vLLM IR]: Batch Invariance Dispatching in vLLM IR

### ISSUE THREAD

**Issue by ProExpertProg** [vllm-project/vllm#40628]
**Related:** #32358
**Status:** Under Discussion
**Written with:** Claude

## Summary

This RFC discusses how batch invariance should be handled within the vLLM IR dispatching model. Currently toggled via VLLM\_BATCH\_INVARIANT=1 and primarily affecting rms\_norm, the question is: **should vLLM IR own batch-invariant dispatching as a first-class concern, or should it be handled outside the IR?**

## Background

The vLLM IR design introduces a unified dispatching and kernel registration mechanism to replace CustomOp. A key motivation for this replacement is that CustomOp dispatching was opaque and hard to reason about. A clear example is that @yewentao256 did not realize that forward\_native (the torch-native implementation) was being used by default in batch-invariant mode (and I didn't know at all what was happening), meaning the batch-invariant triton kernel was not invoked, as it only existed inside forward\_cuda and forward\_hip. Multiple levels of initialization and dispatching made the actual execution path difficult to trace.

Under CustomOp, batch invariance for rms\_norm intersects with three execution paths:

* forward\_cuda / forward\_hip — dispatching between the standard CUDA/HIP custom kernel and batch invariant Triton kernel (also AITER in forward\_hip)  
* forward\_native — used by default during compilation; batch-invariant dispatching is absent here, so it is silently skipped. Instead we relied on Inductor to generate a batch-invariant Triton kernel.

Under vLLM IR, the forward\_\* methods are eliminated and replaced with a single dispatching mechanism. How to represent and route batch-invariant execution within that model is an open question.

---

## Proposed Alternatives

There are basically two decisions embedded in this:

* Do we want to dispatch batch-invariant rms\_norm to native torch ops when compiling the model, or just always use the batch-invariant Triton kernel?  
* If yes (alternatives 1-4), how should this dispatch happen?

Alternatives 1-4 are various ways to dispatch to the native implementation, and alternative 5 just always dispatches to the batch-invariant Triton kernel.

### Option 1: vLLM IR handles batch-invariant dispatching natively _(original plan, [PR \#36816](https://github.com/vllm-project/vllm/pull/36816))_

Batch invariance is a first-class property in the IR. Each implementation declares whether it is batch invariant, and IrOp.dispatch filters implementations accordingly when the batch\_invariant flag is set to true.

### Option 2: Control batch-invariant kernels via vLLM IR priority list _(my new preference)_

Batch invariance is not encoded in the IR itself. Instead, when VLLM\_BATCH\_INVARIANT=1, the KernelConfig initialization logic adjusts the priority list for affected ops to prefer batch-invariant implementations:

```py
class CudaPlatform:
  ...

  @classmethod
  def get_default_ir_op_priority(cls, vllm_config: VllmConfig) -> IrOpPriorityConfig:
    # Native used by default when compiling,
    # use vllm_c kernels where available when no codegen
    cc = vllm_config.compilation_config
    using_inductor = cc.backend == "inductor" and cc.mode != CompilationMode.NONE
    default = ["native"] if using_inductor else ["vllm_c", "native"]

    if using_inductor:
      rms_norm = ["native"]
    elif envs.VLLM_BATCH_INVARIANT:
      rms_norm = ["triton_batch_invariant", "native"]
    else:
      rms_norm = ["vllm_c", "native"]

    return IrOpPriorityConfig.with_default(default, rms_norm=rms_norm)
```

The IR dispatches according to whatever priority list is active and remains unaware of batch invariance as a concept.

### Option 3: Separate rms\_norm\_batch\_invariant IR op

Define a distinct IR op vllm.ir.ops.rms\_norm\_batch\_invariant with its own semantics declaration and implementation registry. Layer code selects which op to invoke based on whether batch invariance is required. This makes the path separate and even visible in the FX graph, but that also means more duplication and manual effort.

### Option 4: Separate dispatching mechanism outside vLLM IR

When VLLM\_BATCH\_INVARIANT=1, execution bypasses IR dispatching entirely and dispatches between the batch-invariant Triton kernel and native implementation directly. The IR is not involved in this path. This does not allow OOT extensibility or custom compile passes but it does at least keep the benefit of native performance, although the native implementation has to be duplicated as we’re not using the IR op (we could manually invoke ir.ops.rms\_norm.impls\[“native”\]).

### Option 5: Batch-invariant path always calls the Triton kernel _(Wentao's preference)_

When batch invariance is required, always dispatch directly to the batch-invariant Triton kernel, with no IR involvement and no priority list manipulation. This matches current behavior on A100 and earlier GPUs, where eager execution is already forced.

B200 perf results for option 5, comparing triton\_batch\_invariant and native implementations (this is before #40408 & #40413 which close roughly half of that gap). Llama was the easiest to demonstrate this on but this applies to Qwen and Deepseek even more as they have more norms per layer (there are some temporary confounding issues right now out of scope of this RFC):

```shell
# triton_batch_invariant
$ VLLM_BATCH_INVARIANT=1 vllm bench latency --model=RedHatAI/Meta-Llama-3.1-8B-Instruct-FP8 --attention-backend=TRITON_ATTN --ir-op-priority.rms_norm=triton_batch_invariant --ir-op-priority.fused_add_rms_norm=triton_batch_invariant -cc.pass_config.fuse_norm_quant=False
Avg latency: 0.43808772719154754 seconds
10% percentile latency: 0.43741614669561385 seconds
25% percentile latency: 0.43829121452290565 seconds
50% percentile latency: 0.4384383929427713 seconds
75% percentile latency: 0.4386132051004097 seconds
90% percentile latency: 0.43903175396844746 seconds
99% percentile latency: 0.43941899626515807 seconds

# native
$ VLLM_BATCH_INVARIANT=1 vllm bench latency --model=RedHatAI/Meta-Llama-3.1-8B-Instruct-FP8 --attention-backend=TRITON_ATTN --ir-op-priority.rms_norm=native --ir-op-priority.fused_add_rms_norm=native -cc.pass_config.fuse_norm_quant=False
Avg latency: 0.3949168597658475 seconds
10% percentile latency: 0.3945025376044214 seconds
25% percentile latency: 0.39466362388338894 seconds
50% percentile latency: 0.3952090779785067 seconds
75% percentile latency: 0.3955157147720456 seconds
90% percentile latency: 0.3957540170755237 seconds
99% percentile latency: 0.3962868914008141 seconds
```

## Discussion

The core trade-off is between **visible, explicit logic in layer/config code** and **a clean, unified IR dispatching model with superior performance**.

Options 4 and 5 are the simplest from an IR perspective: no batch-invariance-related logic enters the core dispatch mechanism. However, both place the batch-invariant path outside the IR entirely, making it a second-class citizen. OOT platforms (that don’t support the Triton kernel) would need to implement their own separate routing logic and override the whole layer rather than using the standard register\_impl interface. 

Option 5 is the simplest, but it carries the classic downsides of a custom kernel: lower performance portability, no portability beyond Triton kernels, no automatic fusion, higher maintenance cost. Option 4 recovers torch.compile\-kernel-gen benefits (including 5% E2E perf) but still lacks OOT extensibility and IR benefits (fusion passes & sequence parallelism/async TP, although those don’t currently work with batch invariance anyway).

Option 3 adds explicit IR visibility but nearly duplicates the rms\_norm op declaration, creates maintenance burden, and requires conditional logic in the RMSNorm layer — complexity that grows if other ops ever need batch-invariant variants.

Option 1 (current plan) keeps the batch-invariant path fully inside the IR, giving it access to unified dispatching, OOT register\_impl, and Inductor fusions. The cost is that the IR dispatch model must understand the concept of batch invariance, adding a layer of reasoning to the core.

Option 2 splits the difference: batch-invariant kernels are still registered through the IR and benefit from unified dispatch and OOT extensibility, but the IR itself has no knowledge of batch invariance — it just follows the active priority list. The batch-invariant override logic lives in CudaPlatform.get\_default\_ir\_op\_priority, which is explicit and easy to audit. The main downside relative to Option 1 is that there is no enforcement that a registered implementation actually satisfies batch invariance — correctness depends on the platform providing the right priority list.

## Recommendation

Option 1 was the original plan and is the most principled long-term design. However, Option 2 is likely a better starting point: it does not preclude migrating to Option 1 in the future, keeps the IR dispatch model slightly simpler while the IR is still stabilizing, and avoids encoding a concept that currently only affects a single op (rms\_norm). No other ops are currently known to require batch-invariant variants, so building first-class IR support for this now is premature. If batch invariance becomes relevant to additional ops, or Inductor fusions on the batch-invariant path prove impactful, migrating to Option 1 remains straightforward.

Feedback from RL post-training users and OOT platform maintainers would be especially valuable.

## Feedback Period.

4/22 - 4/24, this is a high-priority blocker for vLLM IR enablement

## CC List.

@yewentao256 @simon-mo @zhuohan123 @houseroad @tjtanaa @zou3519 @gmagogsfm @angelayi @PaulZhang12 @tlrmchlsmth

**Comment by gmagogsfm** [comment]
Great write up!

In principle, I think option 1 is optimal with cleanest abstraction, batch-invariance is a kernel implementation property, so it should ideally be a flag on IrOpImpl with lowering pass honoring it. But I think it is also reasonable to start with option 2 as it is extremely low cost to get going.

OOC, does batch-variance only affect one op? I thought all layernorm ops and fused variants be affected by it?

**Comment by ProExpertProg** [comment]
Yeah, all norm variants would be affected, but we currently don't have batch-invariant kernels for the fused variants.

**Comment by wxsIcey** [comment]
Thanks for the thorough write-up. I'm sharing my views based on vllm-ascend (OOT platform).

**On Option 2 as a starting point.** I think this is a good idea. The priority list mechanism in `get_default_ir_op_priority()` maps naturally to OOT platforms that have batch-invariant implementation. As @gmagogsfm noted, fused norm variants are equally affected. On Ascend, `torch_npu.npu_add_rms_norm` is non-deterministic due to the fusion itself. The batch-invariant fix is ​​to simply unfuse it: `add` followed by a call to the `torch_npu.rms_norm`(with batch-invariant). The cost of including `fused_add_rms_norm` in scope is minimal, and it would let us eliminate our monkey-patch on `npu_add_rms_norm` entirely.

**On Options 4 and 5.** Option 4 is explicitly OOT-unfriendly as you noted. Option 5 is problematic for platforms like Ascend that have native kernels outperforming Triton.

**On Option 1 as the long-term direction.** As the fused norm discussion illustrates, the number of ops requiring batch-variant will grow. At that point, per-op priority overrides across every platform's priority config become hard to maintain. Having each implementation declare its batch-invariant status at registration time may be better.

cc @wangxiyuan

**Comment by yewentao256** [comment]
Thank you very much for the RFC! It looks good

My main concern is IR related changes involve more complexity and also depends on an implicit mean override to realize batch invariance. `_batch_invariant_LIB.impl("aten::mean.dim", mean_batch_invariant, "CUDA")`

It is also much harder to maintain and debug. We already saw a series of issues and flaky tests when upgrading PyTorch to 2.10.
More importantly, there is a real functional constraint here: we need to force eager on A100 and earlier GPUs. (In this path we automatically downgrade to call the triton kernel)

On the performance side, as [#40413](https://github.com/vllm-project/vllm/pull/40413) landed, I think the perf improvement would be less. We don't really spend the time to improve the performance, if we aim to do so, something like [#40408](https://github.com/vllm-project/vllm/pull/40408) would have more benefit. I don't think it worth the complexity we introduce here.

So overall, my preference is to keep batch invariance outside the IR. That keeps the IR dispatch model simpler and more maintainable, while still leaving room for us to optimize the batch-invariant path separately where it actually matters.

**Comment by ProExpertProg** [comment]
@yewentao256 do you have a proposed solution/workaround for OOT concerns? And for forcing eager on A100, options 1-3 give us an easy way to do this, just override op priority for `rms_norm` to `triton_batch_invariant` - assuming it's really `rms_norm` that is the issue.

**Comment by yewentao256** [comment]
I don't have enough context for OOT platform, could @wxsIcey share more information?

**Comment by gmagogsfm** [comment]
> Thank you very much for the RFC! It looks good
> 
> My main concern is IR related changes involve more complexity and also depends on an implicit mean override to realize batch invariance. `_batch_invariant_LIB.impl("aten::mean.dim", mean_batch_invariant, "CUDA")`
> 
> It is also much harder to maintain and debug. We already saw a series of issues and flaky tests when upgrading PyTorch to 2.10. More importantly, there is a real functional constraint here: we need to force eager on A100 and earlier GPUs. (In this path we automatically downgrade to call the triton kernel)
> 
> On the performance side, as [#40413](https://github.com/vllm-project/vllm/pull/40413) landed, I think the perf improvement would be less. We don't really spend the time to improve the performance, if we aim to do so, something like [#40408](https://github.com/vllm-project/vllm/pull/40408) would have more benefit. I don't think it worth the complexity we introduce here.
> 
> So overall, my preference is to keep batch invariance outside the IR. That keeps the IR dispatch model simpler and more maintainable, while still leaving room for us to optimize the batch-invariant path separately where it actually matters.

Hi @yewentao256 , does Option 5 mean we would only allow one backend (currently Triton) for ops that have batch variance problems?

On the complexity side, my naive thought is that even option 1 isn't too crazily complex, i.e. updating a priority list based on batch invariance. Could you elaborate what your concern about maintainability is? Is it about interactions between different flags/configs? Thanks

**Comment by yewentao256** [comment]
Not binded to Triton, we can also have Cuda kernels specially for batch invariance, just like https://github.com/vllm-project/vllm/pull/40413.

The main concern of the complexity is that this is a deterministic path, which should be separate from IR which is much more dynamic. If someday the implicit mean override or something else breaks, it would be very difficult for us to debug which part breaks the invariance. layer code + IR priority + compile/eager path + PyTorch lowering + ATen override...

**Comment by wxsIcey** [comment]
Explaining some issues with OOT platforms @yewentao256 

Similar to CUDA, the OOT platform has operators specific to its hardware. Therefore, we currently need to rewrite `RMSnorm` and other classes. Luka's IR abstracts away the differences in underlying operators, making it possible for the OOT platform to define its own kernel without rewriting the entire class. However, Option 5 conflicts with this, adding a new maintenance burden. I believe that supporting batch invariance still depends on different operator implementations, and I don't expect to introduce a new branch.

Hopefully, this issue can be considered.

**Comment by yewentao256** [comment]
Thanks for the information!

Is batch invariance widely used in OOT platform? And from my understanding, current cuda and triton kernel can't be applied to OOT as well, is this correct?

I am thinking if OOT platform itself could directly call a native version of RMS norm with `mean` override.

`_batch_invariant_LIB.impl("aten::mean.dim", mean_batch_invariant, "CUDA")`

```bash
def rms_norm(
    x: Tensor, weight: Tensor | None, epsilon: float, variance_size: int | None = None
) -> Tensor:
    orig_dtype = x.dtype
    x = x.to(torch.float32)
    x_var = x if variance_size is None else x[..., :variance_size]
    variance = x_var.pow(2).mean(dim=-1, keepdim=True)
    x = x * torch.rsqrt(variance + epsilon)
    if weight is not None:
        x = x.to(weight.dtype) * weight
    return x.to(orig_dtype)
```

Advanced user can wrap this with `torch.compile` and get speed up, but I don't think this should be supported officially in vLLM.

**Comment by ProExpertProg** [comment]
@yewentao256 yes but the question is, where is the extension point for OOT platform? If the batch invariant path is hardcoded, OOT platform cannot plugin, remember they aren't able to modify source code. So they have to override the whole layer. Meanwhile IR approach is pluggable.

**Comment by yewentao256** [comment]
> they aren't able to modify source code

I don't think this is true for python.

```py
import torch
import vllm.model_executor.layers.batch_invariant as bi
import vllm.model_executor.layers.layernorm as ln

def my_rms_norm(input, weight, eps=1e-6):
    x = input.float()
    var = x.pow(2).mean(dim=-1, keepdim=True)
    y = x * torch.rsqrt(var + eps)
    return y.to(input.dtype) * weight

compiled_my_rms_norm = torch.compile(my_rms_norm)

bi.rms_norm_batch_invariant = compiled_my_rms_norm
ln.rms_norm_batch_invariant = compiled_my_rms_norm
```

This is a workaround only, if it is true that there are a lot of users in OOT using batch invariance, we can also have a small registry-replacement specially for that, not necessirily inside IR. *(one question is that current Matrix mul, log_softmax is triton as well, how does that work with OOT?)

There are always tradeoff between designs

And for your information, when upgrading to pytorch 2.12, the compiled path doesn't work again and they spent a lot of time debuging with the issue. https://github.com/pytorch/pytorch/issues/181248

This might not directly related with our case but it clearly show that we shouldn't make it complicated any more now. Again, if someday the implicit mean override or something else breaks, it would be very difficult for us to debug which part breaks the invariance. layer code + IR priority + compile/eager path + PyTorch lowering + ATen override... We can't bear the effort

**Comment by zou3519** [comment]
@yewentao256 in https://github.com/pytorch/pytorch/issues/181248#issuecomment-4361716019 the problem is not torch.compile. We've discovered the test fails with --enforce-eager=True.
